src/lib.rs

```diff
@@ -14,6 +14,39 @@
 //! * To construct a pool, you must supply a [resolver::Resolver] and
 //!   a [backend::Connector]. These are interfaces which specify "how to find
 //!   backends" and "how to create connections to a backend", respectively.
+//!
+//! # DTrace probes
+//!
+//! qorb contains a number of DTrace USDT probes, which fire as the qorb pool
+//! manages its connections. These probes provide visibility into when qorb
+//! initiates connections; checks health; and vends out claims from a pool. The
+//! full list of probes is:
+//!
+//! - `claim-start`: Fires before attempting to make take a claim from the pool.
+//! - `claim-done`: Fires before returning a successful claim to the client.
+//! - `claim-failed`: Fires on failure to take a claim from the pool.
+//! - `connect-start`: Fires before attempting a connection to a backend.
+//! - `connect-done`: Fires after successfully connecting to a backend.
+//! - `connect-failed`: Fires after failing to connect to a backend.
+//! - `handle-claimed`: Fires after claiming a handle from the pool, before
+//!   returning it to the client.
+//! - `handle-recycled`: Fires when a handle is returned to the pool, after it
+//!   is dropped.
+//! - `health-check-start`: Fires when a health check for a backend starts.
+//! - `health-check-done`: Fires when a health check for a backend completes
+//!   successfully.
+//! - `health-check-failed`: Fires when a health check for a backend fails.
+//! - `recycle-start`: Fires before attempting to recycle a connection.
+//! - `recycle-done`: Fires after successsfully recycling a connection.
+//! - `recycle-failed`: Fires when failing to recycle a connection.
+//!
+//! The existence of the probes is behind the `"probes"` feature, which is
+//! enabled by default. Probes are zero-cost unless they are explicitly enabled,
+//! by tracing the program with the `dtrace(1)` command-line tool.
+//!
+//! Also note that consumers of the `qorb` pool need to call
+//! [`usdt::register_probes`] to register the probes with the OS. `qorb` does
+//! not do so automatically.
```

Comment:
There's a good argument to be made for doing this automatically. That's what we do in `dropshot` for example. The downside is that it changes the `Pool::new()` constructor to be fallible, so it can return a registration error, or we simply swallow that error. Either of those seems fine to me, if that's preferable.

Reply:
hrmmm I'm struggling to have a great opinion here. Looking at omicron, I see tons of calls to `register_probes`, but most of them seem to be unwrapping?

AFAICT the only call to `register_probes` in Omicron that actually returns an error is from `oximeter/db/src/shells/native.rs`, and everyone else seems to be panicking immediately (including other calls in `nexus/db-queries/src/db/pool.rs`)

I'm unclear what work `register_probes` is doing, to infer how bad this actually is. If this is like a "platform support is fundamentally broken, this should always work" Result, then unwrapping doesn't seem *terrible*. On the other hand, if this is something which could happen on e.g. a degraded / overloaded system, it doesn't seem like we should be unwrapping ever, and propagating an error is probably the right call.

My worry with swallowing the error is that this support could break easily, and we could go a long time without noticing.

Reply:
This fails if an `ioctl` to the DTrace kernel module fails. If that happens, then yeah, the system is probably very sad. I've tended to avoid unwrapping in a library like this, because that's something the caller should get to decide how to handle. So in `dropshot`, we return the state of the registration attempt as well.

My suggestion is to do the registration automatically, and keep track of the registration state in the `Pool` one creates. Registration is fairly inexpensive (basically a pretty cheap syscall), and also thread-safe. So we can mostly just do it for clients, but those applications which care about a failure can `assert` that the status is what they expect.

Reply:
> My suggestion is to do the registration automatically, and keep track of the registration state in the Pool one creates.

To make sure I grok this, are you proposing returning a Result, or caching the result inside the Pool struct?

Reply:
The latter, to avoid changing the public API. This could either be the direct return value of `usdt::register_probes()`, or maybe just a bool or simple enum indicating whether registration succeeded.

Reply:
idk, it scares me a bit to start doing operations and leaving the result far away. I'd normally be on-board if the goal of API compatibility was really high, but omicron is the main consumer here - if you're down to update the callers in omicron, I think we should feel free to break the API to handle the error explicitly?

Reply:
I agree API compatibility seems low-ish priority. I'm also hesitant to return a `Result<Pool, Error>` in this case -- that means failing to register the probes will block anything else from working, which seems heavy-handed. We could take the `std::sync::Mutex` approach, I guess, and return the actual pool in either case, and just use whether it was in the `Ok(_)` or `Err(_)` variant to signal whether registration worked. I with your point that putting the error in some random field opens the door to forgetting to check it, so at least this way you have to decide what to do in that case before proceeding.

Reply:
Discussed this a bit in chat with @smklein. This problem is as annoying as when we first built the USDT crate, mostly because there are no great options and a failing registration call is _extremely_ unlikely. But among the shitty tradeoffs, we decided that taking the `std::sync::Mutex` approach is least-terrible. I'm going to change the pool constructor to return a `Result<Pool, E>`, where that error type still lets you get access to the pool if the registration fails. This is pretty similar to keeping the registration result as a field of the pool, but requires the caller to explicitly handle that case. Applications can just unwrap if they want to ensure the probes are registered. Those that don't care can get the pool anyway through the error variant.

Reply:
Done in 393204d